File: src/game/wave-physics/mesh-building/builders/cpu-lagrangian/bounds.ts
import type { TerrainDataForWorker } from "../../MeshBuildTypes";
import type { WaveBounds } from "./types";

/** Number of 32-bit values per contour (must match LandMass.FLOATS_PER_CONTOUR) */
const FLOATS_PER_CONTOUR = 13;

/**
 * Compute a wave-aligned bounding box from root terrain contours.
 * Projects root contour bounding boxes onto the wave direction to find
 * the upwave, downwave, wave-left, and wave-right extents.
 */
export function computeBounds(
  terrain: TerrainDataForWorker,
  wavelength: number,
  waveDx: number,
  waveDy: number,
): WaveBounds {
  const perpDx = -waveDy;
  const perpDy = waveDx;
  const view = new DataView(terrain.contourData);

  let minProj = Infinity;
  let maxProj = -Infinity;
  let minPerp = Infinity;
  let maxPerp = -Infinity;

  for (let ci = 0; ci < terrain.contourCount; ci++) {
    const byteBase = ci * FLOATS_PER_CONTOUR * 4;
    const depth = view.getUint32(byteBase + 16, true);
    if (depth !== 0) continue; // only root contours

    const bMinX = view.getFloat32(byteBase + 32, true);
    const bMinY = view.getFloat32(byteBase + 36, true);
    const bMaxX = view.getFloat32(byteBase + 40, true);
    const bMaxY = view.getFloat32(byteBase + 44, true);

    // Project all 4 bbox corners onto wave direction
    for (const [cx, cy] of [
      [bMinX, bMinY],
      [bMaxX, bMinY],
      [bMaxX, bMaxY],
      [bMinX, bMaxY],
    ]) {
      const proj = cx * waveDx + cy * waveDy;
      const perp = cx * perpDx + cy * perpDy;
      if (proj < minProj) minProj = proj;
      if (proj > maxProj) maxProj = proj;
      if (perp < minPerp) minPerp = perp;
      if (perp > maxPerp) maxPerp = perp;
    }
  }

  if (minProj === Infinity) {
    return { minProj: -500, maxProj: 500, minPerp: -500, maxPerp: 500 };
  }

  const margin = Math.max(2000, wavelength * 20);
  return {
    minProj: minProj - margin,
    maxProj: maxProj + margin,
    minPerp: minPerp - margin,
    maxPerp: maxPerp + margin,
  };
}

File: pipeline/wavemesh-builder/src/bounds.rs
//! Wave-aligned bounding box computation from root terrain contours.
//! Mirrors computeBounds.ts.

use crate::config::MeshBuildBoundsConfig;
use terrain_core::level::{TerrainCPUData, FLOATS_PER_CONTOUR};
use crate::wavefront::{WaveBounds, WaveParams};

/// Compute a wave-aligned bounding box from root terrain contours, with
/// wavelength-scaled margins for upwave, downwave, and crosswave directions.
pub fn compute_bounds(
    terrain: &TerrainCPUData,
    wp: &WaveParams,
    config: &MeshBuildBoundsConfig,
) -> WaveBounds {
    let cd = &terrain.contour_data;

    let mut min_proj = f64::INFINITY;
    let mut max_proj = f64::NEG_INFINITY;
    let mut min_perp = f64::INFINITY;
    let mut max_perp = f64::NEG_INFINITY;

    for ci in 0..terrain.contour_count {
        let base = ci * FLOATS_PER_CONTOUR * 4;
        let b = &cd[base..base + FLOATS_PER_CONTOUR * 4];
        let depth = u32::from_le_bytes([b[16], b[17], b[18], b[19]]);
        if depth != 0 {
            continue;
        } // only root contours

        let b_min_x = f32::from_le_bytes([b[32], b[33], b[34], b[35]]) as f64;
        let b_min_y = f32::from_le_bytes([b[36], b[37], b[38], b[39]]) as f64;
        let b_max_x = f32::from_le_bytes([b[40], b[41], b[42], b[43]]) as f64;
        let b_max_y = f32::from_le_bytes([b[44], b[45], b[46], b[47]]) as f64;

        for &(cx, cy) in &[
            (b_min_x, b_min_y),
            (b_max_x, b_min_y),
            (b_max_x, b_max_y),
            (b_min_x, b_max_y),
        ] {
            let proj = cx * wp.wave_dx + cy * wp.wave_dy;
            let perp = cx * wp.perp_dx + cy * wp.perp_dy;
            if proj < min_proj {
                min_proj = proj;
            }
            if proj > max_proj {
                max_proj = proj;
            }
            if perp < min_perp {
                min_perp = perp;
            }
            if perp > max_perp {
                max_perp = perp;
            }
        }
    }

    if min_proj == f64::INFINITY {
        let half = config.fallback_half_extent_ft;
        return WaveBounds {
            min_proj: -half,
            max_proj: half,
            min_perp: -half,
            max_perp: half,
        };
    }

    let upwave = (wp.wavelength * config.upwave_margin_wavelengths).max(config.min_margin_ft);
    let downwave = (wp.wavelength * config.downwave_margin_wavelengths).max(config.min_margin_ft);
    let crosswave = (wp.wavelength * config.crosswave_margin_wavelengths).max(config.min_margin_ft);

    WaveBounds {
        min_proj: min_proj - upwave,
        max_proj: max_proj + downwave,
        min_perp: min_perp - crosswave,
        max_perp: max_perp + crosswave,
    }
}
